test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/PreserveTypesAndAttributesTests.cs

```diff
@@ -0,0 +1,250 @@
+//---------------------------------------------------------------------
+// <copyright file="PreserveTypesAndAttributesTests.cs" company="Microsoft">
+//      Copyright (C) Microsoft Corporation. All rights reserved. See License.txt in the project root for license information.
+// </copyright>
+//---------------------------------------------------------------------
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace Microsoft.OData.Client.Tests.ALinq
+{
+    /// <summary>
+    /// Tests to check original resource type is used when retrieving attributes like OriginalNameAttribute
+    /// even if the IQueryable<T> object is casted to another type.
+    /// </summary>
+    public class PreserveTypesAndAttributesTests
+    {
+        private readonly TestContext ctx;
+
+        public PreserveTypesAndAttributesTests()
+        {
+            string uri = "https://services.odata.org/V4/Northwind/Northwind.svc";
+            this.ctx = new TestContext(new Uri(uri));
+        }
+
+        [Fact]
+        public void OneLevelCase1()
+        {
+            IQueryable<ICustomer> query1 = ctx.OneLevelCustomers1;
+            var result1 = query1.Where(c => c.CorporationName != "").OrderBy(c => c.Id).Take(5).ToList();
```

Comment:
These tests are making network requests to `https://services.odata.org/V4/Northwind/Northwind.svc`, right? Should we make the test less reliant on the network instead? I think it would be sufficient to test when the URI is generated with the expected property names, without having to make the network requests and return mock responses. Do we have other unit tests doing this @gathogojr  @KenitoInc @ElizabethOkerio?

Reply:
Yes, it relies on external service. I understand what you mean and I think that (for this use case) comparing URIs could be enough. However, other issues, like the one from my other PR (#2863) need to check the response in order to verify that it is working fine.
Could it be an option to add a minimalist local http odata service that we could use for testing purposes?

Reply:
@habbes @joecarl You can and should mock the response. We do this by intercepting `OnMessageCreating` call and returning a mocked response. You can refer to the following for example:
https://github.com/OData/odata.net/blob/bbe2ac74487fb01fe14d97aeca131d0ca89d2452/test/FunctionalTests/Microsoft.OData.Client.Tests/ALinq/DollarApplyTestsBase.cs#L306